Mathlib/Order/Filter/Basic.lean

```diff
@@ -0,0 +1,3090 @@
+  rw [← infᵢ_sets_eq_finite, ← Equiv.plift.surjective.infᵢ_comp]
+  rfl
+#align filter.infi_sets_eq_finite' Filter.infᵢ_sets_eq_finite'
+
+theorem mem_infᵢ_finite {ι : Type _} {f : ι → Filter α} (s) :
+    s ∈ infᵢ f ↔ ∃ t : Finset ι, s ∈ ⨅ i ∈ t, f i :=
+  (Set.ext_iff.1 (infᵢ_sets_eq_finite f) s).trans mem_unionᵢ
+#align filter.mem_infi_finite Filter.mem_infᵢ_finite
+
+theorem mem_infᵢ_finite' {f : ι → Filter α} (s) :
+    s ∈ infᵢ f ↔ ∃ t : Finset (PLift ι), s ∈ ⨅ i ∈ t, f (PLift.down i) :=
+  (Set.ext_iff.1 (infᵢ_sets_eq_finite' f) s).trans mem_unionᵢ
+#align filter.mem_infi_finite' Filter.mem_infᵢ_finite'
+
+@[simp]
+theorem sup_join {f₁ f₂ : Filter (Filter α)} : join f₁ ⊔ join f₂ = join (f₁ ⊔ f₂) :=
+  Filter.ext fun x => by simp only [mem_sup, mem_join]
+#align filter.sup_join Filter.sup_join
+
+@[simp]
+theorem supᵢ_join {ι : Sort w} {f : ι → Filter (Filter α)} : (⨆ x, join (f x)) = join (⨆ x, f x) :=
+  Filter.ext fun x => by simp only [mem_supᵢ, mem_join]
+#align filter.supr_join Filter.supᵢ_join
+
+instance : DistribLattice (Filter α) :=
+  { Filter.instCompleteLatticeFilter with
+    le_sup_inf := by
+      intro x y z s
+      simp only [and_assoc, mem_inf_iff, mem_sup, exists_prop, exists_imp, and_imp]
+      rintro hs t₁ ht₁ t₂ ht₂ rfl
+      exact
+        ⟨t₁, x.sets_of_superset hs (inter_subset_left t₁ t₂), ht₁, t₂,
+          x.sets_of_superset hs (inter_subset_right t₁ t₂), ht₂, rfl⟩ }
+
+-- The dual version does not hold! `Filter α` is not a `CompleteDistribLattice`. -/
+instance : Coframe (Filter α) :=
+  { Filter.instCompleteLatticeFilter with
+    infᵢ_sup_le_sup_infₛ := fun f s t ⟨h₁, h₂⟩ => by
+      rw [infᵢ_subtype']
+      rw [infₛ_eq_infᵢ', infᵢ_sets_eq_finite, mem_unionᵢ] at h₂
+      obtain ⟨u, hu⟩ := h₂
+      rw [← Finset.inf_eq_infᵢ] at hu
+      suffices (⨅ i : s, f ⊔ ↑i) ≤ f ⊔ u.inf fun i => ↑i from this ⟨h₁, hu⟩
+      refine' Finset.induction_on u (le_sup_of_le_right le_top) _
+      rintro ⟨i⟩ u _ ih
+      rw [Finset.inf_insert, sup_inf_left]
+      exact le_inf (infᵢ_le _ _) ih }
+
+theorem mem_infᵢ_finset {s : Finset α} {f : α → Filter β} {t : Set β} :
+    (t ∈ ⨅ a ∈ s, f a) ↔ ∃ p : α → Set β, (∀ a ∈ s, p a ∈ f a) ∧ t = ⋂ a ∈ s, p a := by
+  simp only [← Finset.set_binterᵢ_coe, binterᵢ_eq_interᵢ, infᵢ_subtype']
+  refine' ⟨fun h => _, _⟩
+  · rcases (mem_infᵢ_of_finite _).1 h with ⟨p, hp, rfl⟩
+    refine'
+      ⟨fun a => if h : a ∈ s then p ⟨a, h⟩ else univ, fun a ha => by simpa [ha] using hp ⟨a, ha⟩, _⟩
+    refine' interᵢ_congr_of_surjective id surjective_id _
+    rintro ⟨a, ha⟩
+    simp [ha]
+  · rintro ⟨p, hpf, rfl⟩
+    exact interᵢ_mem.2 fun a => mem_infᵢ_of_mem a (hpf a a.2)
+#align filter.mem_infi_finset Filter.mem_infᵢ_finset
+
+/-- If `f : ι → filter α` is directed, `ι` is not empty, and `∀ i, f i ≠ ⊥`, then `infᵢ f ≠ ⊥`.
+See also `infᵢ_neBot_of_directed` for a version assuming `nonempty α` instead of `nonempty ι`. -/
```

Comment:
```suggestion
/-- If `f : ι → Filter α` is directed, `ι` is not empty, and `∀ i, f i ≠ ⊥`, then `infᵢ f ≠ ⊥`.
See also `infᵢ_neBot_of_directed` for a version assuming `Nonempty α` instead of `Nonempty ι`. -/
```